src/commands/serve.rs

```diff
@@ -0,0 +1,201 @@
+use anyhow::Result;
+use clap::Parser;
+use std::{path::PathBuf, pin::Pin, sync::Arc};
+use wasmtime::component::{Component, InstancePre, Linker};
+use wasmtime::{Engine, Store};
+use wasmtime_cli_flags::CommonOptions;
+use wasmtime_wasi::preview2::{Table, WasiCtx, WasiCtxBuilder, WasiView};
+use wasmtime_wasi_http::{body::HyperOutgoingBody, WasiHttpCtx, WasiHttpView};
+
+struct Host {
+    table: Table,
+    ctx: WasiCtx,
+    http: WasiHttpCtx,
+}
+
+impl Host {
+    fn new() -> Result<Self> {
+        let mut table = Table::new();
+        let ctx = WasiCtxBuilder::new().build(&mut table)?;
+        Ok(Host {
+            table,
+            ctx,
+            http: WasiHttpCtx,
+        })
+    }
+}
+
+impl WasiView for Host {
+    fn table(&self) -> &Table {
+        &self.table
+    }
+
+    fn table_mut(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&self) -> &WasiCtx {
+        &self.ctx
+    }
+
+    fn ctx_mut(&mut self) -> &mut WasiCtx {
+        &mut self.ctx
+    }
+}
+
+impl WasiHttpView for Host {
+    fn table(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&mut self) -> &mut WasiHttpCtx {
+        &mut self.http
+    }
+}
+
+/// Runs a WebAssembly module
+#[derive(Parser)]
+#[structopt(name = "run")]
+pub struct ServeCommand {
+    #[clap(flatten)]
+    common: CommonOptions,
+
+    /// Socket address for the web server to bind to. Defaults to 0.0.0.0:8080.
+    #[clap(long = "addr", value_name = "SOCKADDR")]
+    addr: Option<std::net::SocketAddr>,
+
+    /// The WebAssembly component to run.
+    #[clap(value_name = "WASM", required = true)]
+    component: PathBuf,
+}
```

Comment:
I think it'd be best if this had similar options to `wasmtime run`, for example `--allow-precompiled`, `--dir`, `--env`, and `--profile`. That being said I think it's also totally ok for this to be a follow-up PR too.

One thing that I think should also be done is to share the code that configures WASI, e.g. processing many of the `-S` options. For example in theory `wasi-nn` and such should all be possible to enable too. That would require some refactoring, however, so perhaps best to open an issue and/or file a follow-up PR for that.

Reply:
Yep, agreed. I'll do that in a follow-up PR.

Reply:
https://github.com/bytecodealliance/wasmtime/issues/7105